build.py

```diff
@@ -0,0 +1,357 @@
+#!/usr/bin/env python
+
+# Keep this file compatible with Python 2 -- it's the first file users will run.
+# Even if we end up requiring a different version for the rest of the project,
+# this file should probably not crash and burn.
+
+import argparse
+import errno
+import itertools
+import os
+import subprocess
+import sys
+import traceback
+
+try: import ctypes  # necessary for Windows
+except ImportError: ctypes = None
+
+pip = None  # Will be lazily imported if needed (lazy because the rest of this script should run even on installations without pip)
+
+StringishTypes = (str, bytes) if sys.version_info[0] >= 3 else (basestring, str, unicode)
+
+def pip_main(*args):
+    global pip  # Don't import multiple times
+    if pip is None: import pip
+    main = getattr(pip, 'main', None)
+    if not main:
+        # This is technically unsupported... https://stackoverflow.com/a/50255019
+        # However, calling pip _externally_ for an interpreter that is _already_ running
+        # doesn't seem like any less fragile a scenario!
+        from pip._internal import main
+    return main(*args)
+
+def uniquify(items):
+    result = []
+    seen = set()
+    for item in items:
+        if item not in seen:
+            seen.add(item)
+            result.append(item)
+    return result
+
+def search_path_multi(name, inclusion_predicate=None, env_name='PATH', env_exts='PATHEXT' if sys.platform == 'win32' else None):
+    # Expands a path the same way a shell command would on the current system.
+    # Yields an empty sequence if name is not found.
+    parents = os.environ[env_name].split(os.pathsep)
+    exts = [None] + (os.environ[env_exts].split(os.pathsep) if env_exts is not None and env_exts in os.environ else [])
+    if os.path.basename(name) == name:  # only search if there is no directory component in the path
+        for parent in parents:
+            for ext in exts:
+                name_with_ext = name
+                if ext: name_with_ext += ext.lower() if sys.platform == 'win32' else ext
+                if inclusion_predicate is None or inclusion_predicate(parent, name_with_ext):
+                    path = os.path.join(parent, name_with_ext)
+                    if os.path.lexists(path):
+                        yield path
+    elif os.path.lexists(name):
+        yield name
+
+def search_path(name, *args, **kwargs):
+    result = None
+    for result in search_path_multi(name, *args, **kwargs):
+        break
+    return result
+
+def search_path_or_default(name, *args, **kwargs):
+    result = search_path(name, *args, **kwargs)
+    if result is None: result = name
+    return result
+
+def is_native_windows_or_msys(sys_platform):  # Does NOT return true for WSL, which is considered "Linux"
+    return sys_platform == 'msys' or sys_platform == 'win32'
+
+def make_bash_command_line(sep, *commands, **kwargs):  # Constructs Bash commands that are safely quoted
```

Comment:
We shouldn't need this and instead make sure we are not constructing unsafe commands.